hera_cal/nucal.py

```diff
@@ -1013,3 +1032,516 @@ def project_u_model_comps_on_spec_axis(u_model_comps, spectral_filters):
+        minor_cycle_maxiter : int, default=0
+            Number of minor cycles to perform after each major cycle. Minor cycles are performed by fixing the foreground
+            model and solving for the calibration parameters. This is useful for improving convergence.
+        convergence_criteria : float, default=1e-10
+            Convergence criteria for stopping the optimization. If the difference in loss between two iterations is less than
+            convergence_criteria, the optimization will stop.
+        return_model : bool, default=False
+            If True, the model visibilities will be returned.
+
+        Returns:
+        -------
+        gains : dictionary
+            Dictionary mapping antenna numbers to complex gains. Keys are antenna numbers and values are complex gains.
+        model_parameters : dictionary
+            Dictionary containing the model parameters for each polarization. Keys are polarization strings and values are
+        metadata : dictionary
+            Dictionary containing metadata from the optimization. Contains dictionaries for each polarization with the number of 
+            iterations ("niter") and the loss history of the gradient descent ("loss_history").
+        
+        If return_model:
+            model : DataContainer
+                DataContainer containing the model visibilities. DataContainer is of the form {(ant1, ant2, pol): np.array([Ntimes, Nfreqs])}
+        """
+        # Compute spectral and spatial filters
+        self._compute_filters(
+            freqs=data.freqs, spectral_filter_half_width=spectral_filter_half_width, 
+            spatial_filter_half_width=spatial_filter_half_width, eigenval_cutoff=eigenval_cutoff,
+            umin=umin, umax=umax
+        )
+
+        # Assert that the optimizer is valid
+        assert optimizer_name in OPTIMIZERS, f"Optimizer must be one of {OPTIMIZERS}. Got {optimizer_name}."
+        optimizer = OPTIMIZERS[optimizer_name](learning_rate=learning_rate)
+
+        # Compute the estimates of the model components from the data
+        if spatial_estimate_only:
+            init_model_comps = fit_nucal_foreground_model(
+                data, data_wgts, self.radial_reds, self.spatial_filters, solver=linear_solver, share_fg_model=share_fg_model, 
+                return_model_comps=True, tol=linear_tol
+            )
+            init_model_comps = project_u_model_comps_on_spec_axis(init_model_comps, self.spectral_filters)
+        else:
+            init_model_comps = fit_nucal_foreground_model(
+                data, data_wgts, self.radial_reds, self.spatial_filters, solver=linear_solver, share_fg_model=share_fg_model, 
+                spectral_filters=self.spectral_filters, return_model_comps=True, tol=linear_tol
+            )
+            
+        # Compute idealized baseline vectors from antenna positions and calibration flags
+        idealized_antpos = abscal._get_idealized_antpos(cal_flags, self.antpos, data.pols())
```

Comment:
This can give higher-than-2 dimensional antenna positions (and thus degeneracies). Is that a problem for nucal?

Reply:
No, I don't think that's an issue. We've tested nucal out on greater than 2 dimensional antenna positions.